src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala

```diff
@@ -177,12 +177,19 @@ case class DeltaSource(
 
   private def getStartingOffset(
       limits: Option[AdmissionLimits] = Some(new AdmissionLimits())): Option[Offset] = {
-    val version = deltaLog.snapshot.version
+
+    val (version, isStartingVersion) = getStartingVersion match {
+      case Some(v) => (v._1, false)
+      case None => (deltaLog.snapshot.version, true)
+    }
     if (version < 0) {
       return None
     }
+    if (options.startingVersion.isDefined || options.startingTimestamp.isDefined) {
+      return Some(DeltaSourceOffset(tableId, version, -1, isStartingVersion = false))
+    }
```

Comment:
I don't think this is the right thing to return here. `getStartingOffset` returns the "endOffset" for the first micro-batch. What you were doing in this method before was right

Reply:
@brkyvz when I do that, the test _"startingVersion: start from version > 0"_ doesn't return a result. The wrong offset is returned.

Reply:
The idea is that this method starts from startingVersion index -1, and finds the last available element after that, which should be startingVersion + 1 index -1

Reply:
If that were the case, the user would provide `startingVersion = 0` if they wanted to start at version 1... right?

Reply:
Fixed after I used your code snippet from getBatch